torch/csrc/autograd/input_buffer.cpp

```diff
@@ -129,96 +149,137 @@ static void accumulate(
+  // canonical stream, e.g. this can happen if forward is DtoH
+  const std::optional<c10::Stream>& opt_producer_stream =
+      (opt_producer_stream_.has_value()
+           ? opt_producer_stream_
+           : (is_accelerator
+                  ? std::optional<c10::Stream>(
+                        at::accelerator::getCurrentStream(device->index()))
+                  : std::nullopt));
+  if (is_accelerator) {
+    TORCH_INTERNAL_ASSERT(opt_consumer_stream && opt_producer_stream);
+  }
+
+  // [ Single producer ]
+  // If there's only a single producer, there is no accumulation involved.
+  // All we need to do is:
+  // - (If streams are involved) Have the consumer canonical stream wait for
+  //   the producer canonical stream
+  // - Move var into the buffer.
+  if (num_dependencies == 1) {
+    if (is_accelerator) {
+      mb_wait_stream(opt_consumer_stream, opt_producer_stream, device_type);
+      mb_record_stream_any_impl(
+          var, *opt_consumer_stream, /*prev_stream=*/*opt_producer_stream);
+    }
+    TORCH_INTERNAL_ASSERT(current_index == 0);
+    TORCH_INTERNAL_ASSERT(!buffer[pos].defined())
+    buffer[pos] = std::move(var);
+    return;
+  }
+  // [ First producer ]
+  // In the multiple producer case, we need to handle accumulation.
+  // When we see the very first producer, we do a couple things:
+  // - Determine the accumulation stream:
+  //    case 1)   non-accelerator             -> n/a
+  //    case 2/3) var device matches consumer -> consumer stream
+  //    case 4)   var device matches producer -> producer stream
+  //    case 5)   var device matches neither  -> var device's current stream
+  // - Move var into the buffer.
+  // - In the 2/3 case we also stash the producer stream because the
+  //   accumulation stream would need to sync with it later.
+  //   See test_side_stream_backward_overlap
+  if (current_index == 0) {
+    if (is_accelerator) {
+      // NOLINTNEXTLINE(bugprone-unchecked-optional-access)
+      if (opt_consumer_stream->device() == *device) {
+        // Case 2/3
+        opt_accum_streams[pos] = opt_consumer_stream;
+        opt_first_producer_streams[pos] = opt_producer_stream;
       } else {
-        // (5)
-        opt_accumulate_stream = guard.getDefaultStream(*device);
-      }
-      if (opt_sync_stream && (opt_accumulate_stream != opt_sync_stream)) {
-        // (3b), (4b)
-        c10::OptionalDeviceGuard device_guard{opt_sync_stream->device()};
-        auto event = c10::Event{device_type};
-        event.record(*opt_sync_stream);
-        opt_accumulate_stream->wait(event);
-        const auto guard = c10::impl::VirtualGuardImpl(device_type);
-        record_stream_any_impl(var, *opt_accumulate_stream);
+        // Case 4/5
+        if (opt_producer_stream->device() == *device) {
+          opt_accum_streams[pos] = opt_producer_stream;
+        } else {
+          opt_accum_streams[pos] =
+              at::accelerator::getCurrentStream(device->index());
```

Comment:
So we're going with the current stream on that device.
Can you give some context on why you chose this resolution?